`Adaptive learning`: Improve participation in learning path view

MaximilianAnzinger opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
This PR addresses improvements related to participation in exercises and lecture units using Learning Paths in Artemis.
Stacked on #6747 .

### Description
<!-- Describe your changes in detail -->
- fixes several style issues
- adds generation of LP if a student's LP is missing and requested for the first time

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->
Prerequisites:
- 1 Admin
- 1 Instructor
- 1 Students/User (not enrolled in the course)
- All Exercises and a Lecture with all lecture units excluding exercise unit
- One Competency that contains all the exercises and lecture units

1. Log in to Artemis as Instructor
2. Navigate to Learning Path Management and enable learning paths
3. Log in to Artemis as Admin
4. Manually add the Student/User to the student group of the course
5. The List of Learning Paths in `Learning Path Management` of the course should not contain the Student's Learning Path
6. Log in to Artemis as Student/User
7. Navigate to the Courses Learning Path tab
8. Log in to Artemis as Instructor
9. Navigate to Learning Path Management and make sure you can see the student's learning path in the list
10. Log in to Artemis as Student and navigate to the Courses Learning Path tab
For each exercise/lecture unit
11. Click on the `Next` button to open a lecture unit/exercise
12. Participate in the lecture unit/exercise and make sure it works as expected
13. Also check the conversations tab on the right side (correct channel? Is it possible to post & react?)
14. Return to the learning path and continue (11.) until there are no new recommendations

### Review Progress
<!-- Each Pull Request should be reviewed by at least two other developers. The code, the functionality (= manual test) and the exam mode need to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->
<!-- All PRs that might affect the exam mode (e.g. change a client component that is also used in the exam mode) need an additional verification that the exam mode still works. -->

#### Performance Review
- [ ] I confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2
#### Exam Mode Test
- [ ] Test 1
- [ ] Test 2

### Test Coverage
<!-- Please add the test coverages for all changed files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan. -->
<!-- The line coverage must be above 90% for changes files and you must use extensive and useful assertions for server tests and expect statements for client tests. -->
<!-- Note: Use the table below and confirm in the last column that you have implemented extensive assertions for server tests and expect statements for client tests. -->
<!--       You can use `supporting_script/generate_code_cov_table/generate_code_cov_table.py` to automatically generate one from the corresponding Bamboo build plan artefacts. -->
<!--       Remove rows with only trivial changes from the table. -->
<!--
| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| ExerciseService.java | 85% | ✅                           |
| programming-exercise.component.ts | 95% | ✅              |
-->

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->
<img width="1719" alt="Screenshot 2023-08-01 at 12 55 58" src="https://github.com/ls1intum/Artemis/assets/44003963/ac19fccc-9e81-49dc-8417-070742f369f8">

JohannesStoehr commented on src/main/java/de/tum/in/www1/artemis/web/rest/LearningPathResource.java:
```diff
@@ -110,8 +111,18 @@ public ResponseEntity<Long> getLearningPathId(@PathVariable Long courseId) {
+        // generate learning path if missing
         User user = userRepository.getUser();
-        LearningPath learningPath = learningPathRepository.findByCourseIdAndUserIdElseThrow(course.getId(), user.getId());
+        final var learningPathOptional = learningPathRepository.findByCourseIdAndUserId(course.getId(), user.getId());
```
How big is this the learning path object? Would it make sense to only retrieve the id from the database, since this is the only thing you return?

MaximilianAnzinger replied:
Other than the id it should only contain the progress (`int`).
Adding a proper query to save on one integer seems overkill. I don't even know if this would actually decrease perforce.

MaximilianAnzinger commented:
Testing session backlog:
- breadcrumbs: fixed
- translation: adjusted (also added some other tooltips that might be helpful)
- prev/next-buttons: waiting for meeting with Stephan and Patrick to get feedback on what they envision the UI to look like
- progress: Investigating

JohannesStoehr commented on src/main/java/de/tum/in/www1/artemis/service/LearningPathService.java:
```diff
@@ -86,11 +86,11 @@ public void generateLearningPaths(@NotNull Course course) {
      * @param user   student for which the learning path is generated
      */
-    public void generateLearningPathForUser(@NotNull Course course, @NotNull User user) {
+    public LearningPath generateLearningPathForUser(@NotNull Course course, @NotNull User user) {
```
Codacy

JohannesStoehr commented on src/main/webapp/app/course/learning-paths/learning-paths.module.ts:
```diff
@@ -13,6 +13,7 @@ import { LearningPathGraphNodeComponent } from 'app/course/learning-paths/learni
 import { NgxGraphModule } from '@swimlane/ngx-graph';
 import { ArtemisLectureUnitsModule } from 'app/overview/course-lectures/lecture-units.module';
 import { LearningPathProgressModalComponent } from 'app/course/learning-paths/learning-path-management/learning-path-progress-modal.component';
+import { LearningPathHistoryStorageService } from 'app/course/learning-paths/participate/learning-path-history-storage.service';
```
Unused
